zipline/protocol.py

```diff
@@ -194,6 +200,87 @@ def current_portfolio_weights(self):
         return position_values / self.portfolio_value
 
 
+class AlgorithmPortfolio(Portfolio):
+
+    def __init__(self, data_portal, benchmark=None):
+        super(AlgorithmPortfolio, self).__init__()
+        self.data_portal = data_portal
+        self.benchmark = benchmark
+        self.current_date = None
+
+    @property
+    def expected_shortfall(self):
+        """
+        Function for computing expected shortfall (also known as CVaR, or
+        Conditional Value at Risk) for the portfolio according to the assets
+        currently held and their respective weight in the portfolio.
+
+        This function requires a data portal in order to retrieve price
+        histories of the assets in the portfolio.
+        """
+        data_portal = self.data_portal
+        current_date = self.current_date
+
+        # If we do not even have a year's worth of data to look back on
+        # then the expected shortfall calculation will not be reliable, so
+        # just return NaN. If we only have between one and two years of
+        # data just use what is available.
+        num_days_of_data = data_portal.trading_calendar.session_distance(
+            self.start_date, current_date,
+        )
+        if num_days_of_data < TRADING_DAYS_PER_YEAR:
+            return np.NaN
+        elif num_days_of_data < CVAR_LOOKBACK_DAYS:
+            num_lookback_days = num_days_of_data
+        else:
+            num_lookback_days = CVAR_LOOKBACK_DAYS
+
+        # Series mapping each asset to its portfolio weight.
+        weights = self.current_portfolio_weights
+
+        assets = map(self._asset_for_history_call, weights.index)
+        prices = data_portal.get_history_window(
+           assets=assets,
+           end_dt=current_date,
+           bar_count=num_lookback_days,
+           frequency='1d',
+           field='price',
+           data_frequency='daily',
+        )
+        asset_returns = prices.pct_change()
+
+        return conditional_value_at_risk(
+            returns=asset_returns.fillna(0).dot(weights.values),
+            cutoff=CVAR_CUTOFF,
+        )
+
+    def _asset_for_history_call(self, asset):
+        calendar = self.data_portal.trading_calendar
+        asset_finder = self.data_portal.asset_finder
+        current_date = self.current_date
+
+        if isinstance(asset, Equity):
+            num_days_of_data = calendar.session_distance(
```

Comment:
Correct, as long as you consider a forward filled value as a price point. Otherwise in futures algos we'd have to segregate the history call for equities to use its own calendar.